packages/sources/openexchangerates-test/src/endpoint/forex.ts

```diff
@@ -0,0 +1,103 @@
+import { HttpTransport } from '@chainlink/external-adapter-framework/transports'
+import {
+  PriceEndpoint,
+  priceEndpointInputParameters,
+  PriceEndpointParams,
+} from '@chainlink/external-adapter-framework/adapter'
+import { SingleNumberResultResponse } from '@chainlink/external-adapter-framework/util'
+import { config } from '../config'
+
+interface ResponseSchema {
+  disclaimer: string
+  license: string
+  timestamp: number
+  base: string
+  rates: {
+    [key: string]: number
+  }
+}
+
+export type ForexEndpointTypes = {
+  Request: {
+    Params: PriceEndpointParams
+  }
+  Response: SingleNumberResultResponse
+  Settings: typeof config.settings
+  Provider: {
+    RequestBody: never
+    ResponseBody: ResponseSchema
+  }
+}
+
+const getMappedSymbols = (requestParams: PriceEndpointParams[]) => {
+  const symbolGroupMap: Record<string, { params: PriceEndpointParams[]; base: string }> = {}
+  requestParams.forEach((param) => {
+    const base = param.base.toUpperCase()
+
+    if (!symbolGroupMap[base]) {
+      symbolGroupMap[base] = {
+        base,
+        params: [],
+      }
+    }
+
+    if (!symbolGroupMap[base].params) {
+      symbolGroupMap[base].params = [param]
+    } else {
+      symbolGroupMap[base].params.push(param)
+    }
+  })
+
+  return symbolGroupMap
+}
+
+export const batchTransport = new HttpTransport<ForexEndpointTypes>({
+  prepareRequests: (params, config) => {
+    // OpenExchangeRates supports batching only for base params, so we are grouping params by bases meaning we will send N number of requests to DP where the N is number of unique bases
+    const groupedSymbols = getMappedSymbols(params)
+    return Object.values(groupedSymbols).map((group) => {
+      const { base } = group
+      return {
+        params: group.params,
+        request: {
+          url: 'latest.json',
+          baseURL: config.API_ENDPOINT,
+          params: {
+            app_id: config.API_KEY,
+            base,
+          },
+        },
+      }
+    })
+  },
+  parseResponse: (params, res) => {
+    return params.map((param) => {
+      const result = res.data.rates[param.quote.toUpperCase()]
```

Comment:
Just for safety could we check if `rates` exists first and return a generic error if it doesn't for all params?

Reply:
done